src/Psalm/Internal/Type/Comparator/UnionTypeComparator.php

```diff
@@ -175,7 +175,7 @@ public static function isContainedBy(
                     if ($container_required_param_count > $input_all_param_count
                         || $container_all_param_count < $input_required_param_count
                     ) {
-                        return false;
+                        continue;
```

Comment:
Seems fundamentally wrong to me. You basically removed the only thing in the loop that made it not useless (if I'm not clear, I think you could remove the loop altogether and get the same effect)

I think you need to inverse the logic of the loop. Instead of failing (returning false) at the first issue, you need to return false if and only if every part of the loop is wrong

Reply:
Hmmm, this logic seems correct to me, because we're basically doing

```
foreach ($input_type as $input_type_part) {
    $type_match_found = false;
    foreach ($container_type as $container_type_part) {
        if (<container_type_part contains input_type_part>) {
            $type_match_found = true;
        }
    }
    if (!$type_match_found) {
        return false;
    }
}

return true;
```

There's no place for a `return false` in the second foreach, since we're only acting on the `$type_match_found` variable to signal whether a match was found or not.

Reply:
Sorry, I misread where the loop ended. My previous remark don't make any sense

Reply:
Yeah the loop is huge, it was a bit hard for me too :P